x/staking/keeper/delegation.go

```diff
@@ -643,9 +643,22 @@ func (k Keeper) Delegate(
 	// validator.Tokens
 	lastTotalPower := k.GetLastTotalPower(ctx)
 	maxVotingPowerEnforcementThreshold := k.MaxVotingPowerEnforcementThreshold(ctx)
-	if lastTotalPower.GTE(maxVotingPowerEnforcementThreshold) {
-		newVotingPowerRatio := sdk.NewDecFromInt(validator.Tokens.Add(bondAmt)).Quo(sdk.NewDecFromInt(lastTotalPower))
-		if newVotingPowerRatio.GT(k.MaxVotingPowerRatio(ctx)) {
+
+	// 1 power = Bond Amount / Power Reduction
+	validatorAddtionalPower := bondAmt.Quo(k.PowerReduction(ctx))
+	newTotalPower := lastTotalPower.Add(validatorAddtionalPower)
+
+	// If it's beyond genesis then enforce power ratio per validator if there's more than maxVotingPowerEnforcementThreshold
+	if newTotalPower.GTE(maxVotingPowerEnforcementThreshold) && ctx.BlockHeight() > 0 {
+		// Convert bond amount to power first
+		validatorNewTotalPower := validator.Tokens.Add(bondAmt).Quo(k.PowerReduction(ctx))
+		// Validator's new total power cannot exceed the max power ratio that's allowed
+		newVotingPowerRatio := validatorNewTotalPower.ToDec().Quo(newTotalPower.ToDec())
```

Comment:
how is this expected to behave if for example I send a TX with multiple messages that delegates to ALL validators (one by one in each message) such that the intended end of transaction is for them to still be even, but for example after the first message is executed the voting power would be imbalanced? The transaction would fail, right? is that intended behavior?

Reply:
Yeah, the whole TX would fail. 

If a message is invalid then the TX itself is invalid and should fail right? If TXs aren't atomic then it makes the state transition non-deterministic for the user